packages/socket/lib/cdp-socket.ts

```diff
@@ -0,0 +1,180 @@
+    this._fullNamespace = `${path}${namespace}`
+  }
+
+  async attachCDPClient (cdpClient: CDPClient): Promise<void> {
+    this._cdpSocket = await CDPSocket.init(cdpClient, this._fullNamespace)
+
+    await Promise.all(Object.values(this._namespaceMap).map(async (server) => {
+      return server.attachCDPClient(cdpClient)
+    }))
+
+    super.emit('connection', this._cdpSocket)
+  }
+
+  emit = (event: string, ...args: any[]) => {
+    this._cdpSocket?.emit(event, ...args)
+
+    return true
+  }
+
+  of (namespace: string): CDPSocketServer {
+    const fullNamespace = `${this._path}${namespace}`
+
+    if (!this._namespaceMap[fullNamespace]) {
+      this._namespaceMap[fullNamespace] = new CDPSocketServer({ path: this._path, namespace })
+    }
+
+    return this._namespaceMap[fullNamespace]
+  }
+
+  to (room: string): CDPSocketServer {
+    return this
+  }
+
+  // TODO: figure out end lifecycle/disconnects/etc.
+  close (): void {
+    this._cdpSocket?.close()
+    this.removeAllListeners()
+    this._cdpSocket = undefined
+
+    Object.values(this._namespaceMap).forEach((server) => {
+      server.close()
+    })
+  }
+
+  // TODO: figure out end lifecycle/disconnects/etc.
+  disconnectSockets (close?: boolean): void {
+    this._cdpSocket?.close()
+    this.removeAllListeners()
+    this._cdpSocket = undefined
+
+    Object.values(this._namespaceMap).forEach((server) => {
+      server.disconnectSockets()
+    })
+  }
+}
+
+export class CDPSocket extends EventEmitter {
+  private _cdpClient?: CDPClient
+  private _namespace: string
+  private _executionContextId?: number
+
+  constructor (cdpClient: CDPClient, namespace: string) {
+    super()
+
+    this._cdpClient = cdpClient
+    this._namespace = namespace
+
+    this._cdpClient.on('Runtime.bindingCalled', this.processCDPRuntimeBinding)
+  }
+
+  static async init (cdpClient: CDPClient, namespace: string): Promise<CDPSocket> {
+    await cdpClient.send('Runtime.enable')
+
+    await cdpClient.send('Runtime.addBinding', {
+      name: `cypressSendToServer-${namespace}`,
+    })
+
+    return new CDPSocket(cdpClient, namespace)
+  }
+
+  join = (room: string): void => {
+    return
+  }
+
+  emit = (event: string, ...args: any[]) => {
+    // Generate a unique callback event name
+    const uuid = randomUUID()
+    const callbackEvent = `${event}-${uuid}`
+    let callback: any
+
+    if (typeof args[args.length - 1] === 'function') {
+      callback = args.pop()
+    }
+
+    if (callback) {
+      this.once(callbackEvent, () => {
+        debugVerbose('callback called from browser')
+        callback()
+      })
+    }
+
+    encode([event, callbackEvent, args], this._namespace).then((encoded: any) => {
+      const expression = `
+        if (window['cypressSocket-${this._namespace}'] && window['cypressSocket-${this._namespace}'].send) {
+          window['cypressSocket-${this._namespace}'].send('${JSON.stringify(encoded).replaceAll('\\', '\\\\').replaceAll('\'', '\\\'')}')
```

Comment:
Wondering aloud if there is a better way to do this... I did this originally and it seems to work... but it sure is ugly.

Reply:
I tried a few things and couldn't find a way. Will keep thinking about it though.